api/envoy/api/v2/core/base.proto

```diff
@@ -79,6 +79,13 @@ message Node {
   // version of Envoy is being tested in a heterogeneous fleet. This will be set
   // by Envoy in management server RPCs.
   string build_version = 5;
+
+  // A list of client capabilities indicating extensions that can be provided.
+  // Strings must be in reverse DNS notation. The following values are known:
+  //
+  // "envoy.grpc_client": Indicates that the client is a gRPC client and not a proxy. For
+  //     LDS, the server is required to provide exactly one HTTP API listener.
```

Comment:
This is getting into the gray area of making this user agent specific. I rather we either user the user agent here or try and have this capability be more abstract, e.g. "envoy.listener.{supported listener type}", etc. I.e. describe what the client can do, not what it is.

Reply:
+1. If we don't make it capability based (even if via somewhat opaque strings) I wonder if we really have to specify this at all here? Given that we plan on moving this out into UDPA ultimately, couldn't the fields here be specified separately in relevant products and systems?

Reply:
nit: It might have been discussed already but having a client under `envoy` namespace seems little confusing as envoy is not at all in the picture. Should it be generally called as `xds.client.grpc_client` or if it ultimates moves to UDPA, `udpa.client.grpc_client` or some thing?

Reply:
I agree that this seems messy.  In particular, I am not very happy with the gRPC-specific semantic of the proposed capability, since it is very close to matching on user-agent, and I really don't think we want the management server to be in the business of magically knowing what each client can do.  But at the same time, the main purpose of this capability is to express something fundamental about the client, which is the way that it finds the relevant listener, and I think gRPC may wind up doing this differently than other clients do.

I think the core question here is what the LDS request will look like for gRPC clients.  We've looked at two basic approaches:

1. The LDS request will have its `resource_names` field set to the virtual host (and probably port -- see the comment thread in lds.proto for more discussion) requested by the client application, and the management server would return exactly one listener whose name matches.
2. The LDS request will set this capability and will _not_ populate the `resource_names` field.  The management server will know that it must send the client a single HTTP API listener.

The reason that we tried to avoid approach (1) is that it imposes some constraints on the xDS server:

- The xDS server must provide one listener for each virtual host.
- The listeners' names must match the virtual hosts.

I would prefer to avoid baking these kinds of assumptions into gRPC's use of xDS, since that may make it more complicated for existing management servers to add support for gRPC.

Matt, have you or others given any thought to these issues in the context of Envoy Mobile?

Reply:
> Matt, have you or others given any thought to these issues in the context of Envoy Mobile?

TBH not yet, as I don't think we would dynamically deliver LDS for Envoy Mobile for quite some time. I think it's much more likely that we will build a static set of listeners based on the library configuration, with potentially other APIs fetched from an xDS server. cc @goaway @junr03

Reply:
I've removed the capability proposal from this PR, since we no longer need it for this functionality.